// dts-builder/dist/lib/regexp-patterns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// example: ///<reference path="myfile.d.ts" />
exports.reference = /^\/\/\/\s+<\s*reference\s+path\s*=\s*["'].+["']\s*\/>\s*$/gm;
// example: import *, { foo } from "./bar";
exports.externalModule = /^(\s*import\s+[\*\{}\w\s\n, ]*['"].+['"];)$/gm;
// example: export * from "./lib/index";
exports.externalReExports = /^\s*export\s+[\*\{}\w\s\n, ]+\s+from\s['"].+['"];$/gm;
// Catching all: import ...;
exports.internalModule = /^(\s*import\s+.*;)$/gm;
// example: import("./bar").Foo;
exports.internalInlineModule = /\b(import\(.*)\./gm;
// example: import foo = barbar.thisIsFoo;
exports.internalModuleParts = /^import\s+([a-zA-Z]+)\s*=\s*([a-zA-Z\.0-9_]+);$/;
// example: declare var _default: string; export defaukt _default;
exports.defaultExport = /^\s*declare\s+var\s+_default\s*:\s*string\s*;\s*export\s+default\s+_default\s*;\s*$/gm;
// example: export default ...;
exports.defaultsFromNs = /^\s*export\s+(default\s+)?[a-zA-Z][a-zA-Z0-9_]*\s*;\s*$/gm;
// example: export { foo, bar };
exports.exportDeclarations = /^\s*export\s+{[^}]*};$/gm;
